### PR TITLE
Move uuid into namespace to decouple it from authkey

### PR DESCRIPTION
**Pull request by mhenke1**
This PR makes the uuid an explicit part of a namespace  object containing name and uuid of a namespace.

## Description
In many places the uuid stored in the authkey is now used as the unique way to identify
a namespace. This behavior entangles the authorization mechanisms
with other unrelated parts of the code, making changes and extensions of the authentication mechanism unnecessarily hard.

This PR makes the uuid a member of a new namespace object containing the name and the
uuid of the namespace. All references to the uuid that are not used on the context of authorization are using the new namespace object as source.

[PG3:2375 succceded]

## Related issue and scope
N/A

## My changes affect the following components>
- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/entity/Identity.scala [hunk 1]
```diff
@@ -38,13 +37,17 @@ object UserLimits extends DefaultJsonProtocol {
...
 
+protected[core] case class Namespace(name: EntityName, uuid: UUID)
+
+protected[core] object Namespace extends DefaultJsonProtocol {
+  implicit val serdes = jsonFormat2(Namespace.apply)
+}
+
 protected[core] case class Identity(subject: Subject,
-                                    namespace: EntityName,
+                                    namespace: Namespace,
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Identity.scala]: By having two uuids you allow for them to be different. Is that intended?

File: common/scala/src/main/scala/whisk/core/entity/Identity.scala [hunk 2]
```diff
@@ -38,13 +37,17 @@ object UserLimits extends DefaultJsonProtocol {
   implicit val serdes = jsonFormat3(UserLimits.apply)
 }
 
+protected[core] case class Namespace(name: EntityName, uuid: UUID)
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Identity.scala]: Do we need the nesting vs: add uuid to Identity and replace authkey with secret then?

File: common/scala/src/main/scala/whisk/core/entity/Identity.scala (hunk 1 above)
Comment (mhenke1) [reply common/scala/src/main/scala/whisk/core/entity/Identity.scala]: The PR has to two motivations.
The first one is to make the (mis-)usage of the uuid as unique namespace alias explicit. 
The change  makes unrelated parts of the system (like for example authentication and user metrics)  also unrelated on the code level.  This could of course be done by providing new access functions to the Identity object. 
My second - not fully outspoken -  motivation for the change is to pave the path
to make the authentication extensible and/or exchangeable. For both cases it is imaginable 
the another authentication implementation  might not depend (fully) on the subject db but (additional) on an external identity and access management system that will not provide a uuid.

I plan to discuss a proposal for extending authentication and authorization on the dev list in the very near future. Since I saw this change more as driven by hygienic reasons (see my first motivation) I put it out separately.

File: common/scala/src/main/scala/whisk/core/entity/Identity.scala (hunk 2 above)
Comment (mhenke1) [reply common/scala/src/main/scala/whisk/core/entity/Identity.scala]: please see my comment above

**Comment by rabbah** [comment]
Thanks for the explanation and the context is more clear now. Definitely bring this to the dev list as there is broader interest wrt IAM integration. 

The PR also suggests the promotion of the namespace type more first class rather than an alias for EntityName.

File: common/scala/src/main/scala/whisk/core/entity/Identity.scala [hunk 3]
```diff
@@ -38,13 +37,17 @@ object UserLimits extends DefaultJsonProtocol {
...
+  implicit val serdes = jsonFormat2(Namespace.apply)
+}
+
 protected[core] case class Identity(subject: Subject,
-                                    namespace: EntityName,
+                                    namespace: Namespace,
                                     authkey: AuthKey,
                                     rights: Set[Privilege],
-                                    limits: UserLimits = UserLimits()) {
-  def uuid = authkey.uuid
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Identity.scala]: This method could stay then and refers to namespace.uuid as a slight convenience.

File: common/scala/src/main/scala/whisk/core/entity/WhiskAuth.scala [hunk 4]
```diff
@@ -25,18 +25,18 @@ import scala.util.Try
...
     def write(w: WhiskNamespace) =
-      JsObject("name" -> w.name.toJson, "uuid" -> w.authkey.uuid.toJson, "key" -> w.authkey.key.toJson)
+      JsObject("name" -> w.namespace.name.toJson, "uuid" -> w.namespace.uuid.toJson, "key" -> w.authkey.key.toJson)
 
     def read(value: JsValue) =
       Try {
         value.asJsObject.getFields("name", "uuid", "key") match {
           case Seq(JsString(n), JsString(u), JsString(k)) =>
-            WhiskNamespace(EntityName(n), AuthKey(UUID(u), Secret(k)))
+            WhiskNamespace(Namespace(EntityName(n), UUID(u)), AuthKey(UUID(u), Secret(k)))
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/WhiskAuth.scala]: this implies a database schema change in the database.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 5]
```diff
@@ -187,7 +187,7 @@ trait WhiskActionsApi extends WhiskCollectionAPI with PostActionActivation with
   override def create(user: Identity, entityName: FullyQualifiedEntityName)(implicit transid: TransactionId) = {
     parameter('overwrite ? false) { overwrite =>
       entity(as[WhiskActionPut]) { content =>
-        val request = content.resolve(user.namespace)
+        val request = content.resolve(user.namespace.name)
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: should we instead change these method signature to accept the new Namespace type instead?

File: tests/src/test/scala/whisk/core/limits/ActionLimitsTests.scala [hunk 6]
```diff
@@ -245,7 +245,7 @@ class ActionLimitsTests extends TestHelpers with WskTestHelpers {
     val allowedSize = ActivationEntityLimit.MAX_ACTIVATION_ENTITY_LIMIT.toBytes
 
     // Needs some bytes grace since activation message is not only the payload.
-    val args = Map("p" -> ("a" * (allowedSize - 700).toInt).toJson)
+    val args = Map("p" -> ("a" * (allowedSize - 750).toInt).toJson)
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/limits/ActionLimitsTests.scala]: ? this change seems unrelated

Comment (mhenke1) [reply tests/src/test/scala/whisk/core/limits/ActionLimitsTests.scala]: Not really. Serializing the uuid as an additional part of the namespace has added some crucial bytes to the message so that the "grace bytes" where not enough to get the message passed.

File: common/scala/src/main/scala/whisk/core/entity/Identity.scala (hunk 3 above)
Comment (mhenke1) [reply common/scala/src/main/scala/whisk/core/entity/Identity.scala]: I was asked by Markus to remove that fuction, to make more explicit that the uuid is retrieved from the namespace.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala (hunk 5 above)
Comment (mhenke1) [reply core/controller/src/main/scala/whisk/core/controller/Actions.scala]: Addressed in latest commit
